### DISCUSSION TITLE
怎么设置全局变量文件 并且打包后暴露出这个文件，这个文件不进行编译在TS中报错 比如设置 window.ENV

### DISCUSSION THREAD

**Discussion by shxinlei** [umijs/umi#6153] [Q&A]
比如建立一个文件叫做ENV.js 在打包的时候移入到打包文件中

但是在TS中出现使用  ENV.xxx 的时候出现window 上找不到ENV 的错误 请求结局方案

**Comment by zhtangsh** [comment]
不清楚umijs是怎么做的，我想说一下我的解决方法..
# Coding
假设配置文件如下
```javascript
// src/env.js
const ENV = {
  a: 1,
  b: 2
}
```
在模板`document.js`中引入外部脚本
```html
...
<head>
    <meta charset="utf-8"/>
    <script type="text/javascript" src="<%= context.publicPath %>env.js"></script>
</head>
...
```
在`typing.d.ts`中定义
```typescript
//typing.d.ts
declare const ENV: { a: number, b: number };
```
在`.umirc.ts`中定义
```
import { defineConfig } from 'umi';

export default defineConfig({
  copy: [{ from: 'src/env.js', to: 'env.js' }],
});
```
然后在代码的任何地方就可以直接使用`ENV.a`和`ENV.b`了

# Deploy
在打包文件的`dist/`下放入`env.js`文件，`env.js`的内容可以用envsubst在容器化部署时通过环境变量替换
在`nginx.conf`中加入对应配置
```
    location ~ env\.js {
      alias /var/static/dist/env.js;
      expires -1;
    }
```
我一般打包成docker镜像，这样将同一份代码部署到不同的地方去，相关配置使用系统的ENV来更改

# Bottom line
这个做法有点不那么umi，但是可以解决问题，也希望知道umi自家是如何做的....搜了很久...
